ddclient.in

```diff
@@ -2145,6 +2156,157 @@ sub geturl {
     return $reply;
 }
 
+######################################################################
+## geturl -- via curl
+######################################################################
+sub curl_cmd {
+    my @params = @_;
+    my $tmpfile;
+    my %curl_codes = (  ## Subset of error codes from https://curl.haxx.se/docs/manpage.html
+         3 => "URL malformed. The syntax was not correct",
+         5 => "Couldn't resolve proxy. The given proxy host could not be resolved.",
+         6 => "Couldn't resolve host. The given remote host was not resolved.",
+         7 => "Failed to connect to host.",
+        22 => "HTTP page not retrieved. The requested url was not found or returned another error.",
+        28 => "Operation timeout. The specified time-out period was reached according to the conditions.",
+        35 => "SSL connect error. The SSL handshaking failed.",
+        47 => "Too many redirects. When following redirects, curl hit the maximum amount.",
+        51 => "The peer's SSL certificate or SSH MD5 fingerprint was not OK.",
+        58 => "Problem with the local certificate.",
+        60 => "Peer certificate cannot be authenticated with known CA certificates.",
+        67 => "The user name, password, or similar was not accepted and curl failed to log in.",
+        77 => "Problem with reading the SSL CA cert (path? access rights?).",
+        78 => "The resource referenced in the URL does not exist.",
+    );
+
+    return '' if (scalar(@params) == 0); ## no parameters provided
+
+    # debug("CURL:\r\n%s\r\n",$params); ## Uncomment only if required as may expose credentials
+
+    eval "no warnings 'all'; use File::Temp;";
+    if (!$@) {
+        ## Use system File::Temp to generate filename...
+        $tmpfile = File::Temp->new(TMPDIR => 1,  ## in system temp dir
+                                   UNLINK => 0,  ## we will delete it ourselves
+                                 TEMPLATE => 'ddclient_XXXXXXXXXX');
+        ## File::Temp->new() automatically opens the file for us.
+    } else {
+        ## System does not have File::Temp so generate ourselves...
+        eval "no warnings 'all'; use File::Spec;";
+        $tmpfile = ((!$@) ? File::Spec->tmpdir() : '/tmp').'/ddclient_';
+        my @chars = ("A".."Z", "a".."z", "0".."9");
+        $tmpfile .= $chars[rand @chars] for 1..10;
+        open($tmpfile, '>', $tmpfile) or fatal("Falied to open $tmpfile for curl");
+    }
+    chmod(0600,$tmpfile);
+    debug("CURL Tempfile: %s",$tmpfile);
+    local $\ = "\n";  ## Terminate the file,
+    local $, = "\n";  ## and each parameter, with a newline.
+    print($tmpfile @params);
+    close($tmpfile);
+
+    ## Make sure to delete temp file on Ctrl-C or SIGTERM
+    local $SIG{'INT'} = sub { unlink($tmpfile); exit 2 };
+    local $SIG{'TERM'} = sub { unlink($tmpfile); exit 15 };
```

Comment:
Took me at least 10 minutes to find this comment. GitHub grouped them all together in the mail and I couldn't open this specific...

The file definitely needs to be cleaned up and Multiple Ctrl+C is usually annoying...

Reply:
Current code should handle this properly, no multiple Ctrl-C required.

Reply:
> Took me at least 10 minutes to find this comment.

GitHub's review UI is the worst. I wonder if GitLab has a better review UI.

> Thoughts?

It's hard to get signals right, and the lack of good documentation doesn't help.

For now I wouldn't worry about cleaning up the temporary files if there's a signal. It's not ideal, but the consequences are minor: ddclient is likely to only be terminated once per system reboot, and reboots should clean out stale temporary files. And the probability that ddclient is terminated while curl is running is quite low. Incorrectly handling signals can cause more problems than leaving the occasional temporary file.

Eventually I think we should redo signal handling. Maybe something like [sigtrap](https://perldoc.perl.org/sigtrap.html) combined with registered closures that are executed in [`END` blocks](https://perldoc.perl.org/perlmod.html#BEGIN%2c-UNITCHECK%2c-CHECK%2c-INIT-and-END).  There are also a bunch of scope-based cleanup modules that we might be able to use:
  * [Scope::Cleanup](https://metacpan.org/pod/Scope::Cleanup)
  * [Scope::OnExit](https://metacpan.org/pod/Scope::OnExit)
  * [Scope::Guard](https://metacpan.org/pod/Scope::Guard)
  * [Guard](https://metacpan.org/pod/Guard)
  * [Sub::ScopeFinalizer](https://metacpan.org/pod/Sub::ScopeFinalizer)
  * [Hook::Scope](https://metacpan.org/pod/Hook::Scope)
  * [B::Hooks::EndOfScope](https://metacpan.org/pod/B::Hooks::EndOfScope)
  * [End](https://metacpan.org/pod/End)
  * [Object::Destroyer](https://metacpan.org/pod/Object::Destroyer)
  * [Perl::AtEndOfScope](https://metacpan.org/pod/Perl::AtEndOfScope)
  * [ReleaseAction](https://metacpan.org/pod/ReleaseAction)
  * [Scope::local_OnExit](https://metacpan.org/pod/Scope::local_OnExit)
  * [Value::Canary](https://metacpan.org/pod/Value::Canary)
  * probably several others

I don't know if any are suitable. I suspect most of them are not packaged, and many are C-based. It probably wouldn't be hard to write our own via object destructor.

Reply:
I don't like the idea of knowingly leaving the temp file lying around.  Of all the security issues we have been discussing this is as bad as they come as we knowingly leave an attack vector which could expose credentials.  The current code installs signal handlers, it is very easy to add local signal handlers just for this subroutine and mimic exactly what the current ones do while also deleting the temp file, and if not running as a daemon also exit.  Note that this has nothing to do with how the temp file is created... File::Temp has the same exposure.

Reply:
> I don't like the idea of knowingly leaving the temp file lying around. Of all the security issues we have been discussing this is as bad as they come as we knowingly leave an attack vector which could expose credentials.

How is it a problem? The file is owned by the ddclient user with 0600 permissions. That's the same as `/etc/ddclient.conf`, which also has the sensitive credentials.

> The current code installs signal handlers, it is very easy to add local signal handlers just for this subroutine and mimic exactly what the current ones do while also deleting the temp file, and if not running as a daemon also exit.

I just realized that we don't need to do anything special for signals—the existing signal handlers are sufficient. If a signal arrives while curl is running the existing signal handler will set the appropriate `$caught_*` variable, then normal execution will resume. Curl will have a chance to complete, as will `geturl`.

Reply:
> I just realized that we don't need to do anything special for signals—the existing signal handlers are sufficient. If a signal arrives while curl is running the existing signal handler will set the appropriate `$caught_*` variable, then normal execution will resume. Curl will have a chance to complete, as will `geturl`.

Well that is not how it is working. SIGINT is interrupting curl and no file deletion takes place.  If running in a daemon then setting of `$caught_int` or `$caught_term` breaks out of the while loop and program ends.

Reply:
Huh, that's unexpected. Can you tell me how you tested it so that I can reproduce?

Reply:
Ah, so that's interesting.  I have found the reason why.  To reproduce do the following...
`ddclient -query`
Then hit Ctrl-C as it is displaying the IP addresses.  Now the reason why is because the call to `test_possible_ip` is done immediately before the signal handlers are set.  So we're both right.  If I move the test_possible_ip call to after the signal handlers then you are right, Ctrl-C does nothing (at least while using -query flag).

So in fact adding my signal handlers changes behavior for the non-deamon state allowing the program to exit.  I need to do further testing in the non -query states.

Reply:
Signal handlers have been removed.